Python+ensurepip fails because `--with-ensurepip=upgrade` tries to remove system setuptools, pip

### Steps to reproduce

With current develop, and spack > https://github.com/spack/spack/commit/a2181e9d253d7e7efae411b52ee995023090edd0 by ~~@alalazo~~ @adamjstewart 
```console
spack install python+ensurepip
```

### Error message

Full log: [spack-build-out.txt](https://github.com/spack/spack/files/7918957/spack-build-out.txt)

```console
11:46:15 wdconinc@menelaos ~ $ spack install python+ensurepip 
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/libiconv-1.16-kpfapbvtfaz22jywdasty3tpwqzxpqru
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/libmd-1.0.3-zu7mcmi75uq4wgeuffg43kvpcgkpkr34
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/xz-5.2.5-lpdrhvdwae23m52ovapt6dq7ruqjf6vz
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/zlib-1.2.11-4gznovrgfc7lht33lqctihlhc7qwi25s
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/libffi-3.3-36g2em6gyl3wkfqqt5ccx7fclheurqps
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/berkeley-db-18.1.40-j3b2docyclx7tdpnhetyacu6fe2bfgdr
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/tar-1.34-plve6dwhgyk25fkb5t2shurpoyx3cklz
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/diffutils-3.8-y4o5hwxld7ligawv6zuwlnxkgu5qqu5s
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/libbsd-0.11.3-k6bqxnjkl33mqidtlwm75w2dkw6gidco
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/ncurses-6.2-zutufy4bqs4nwwkwzhh3hycuhdklilku
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/util-linux-uuid-2.36.2-q7lq3hjunyjhymxpfzyw5qcfisogwj5n
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/libxml2-2.9.12-i4vhktmtyzxpr7me3mvskybnylfatwmx
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/bzip2-1.0.8-wqvlh3tyjpd2dekgoxahsaz2fnc3mio5
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/expat-2.4.3-cbpalrovu2346mrgrwpayeyco2uue4hg
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/readline-8.1-4d3y5gvise76nitctkulcvph6sn4ootb
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/gettext-0.21-zslfuolpfwoxvt4hfxulaz233yhwvpul
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/sqlite-3.37.1-cvhqf7iqwv5dbozowabj3amxva7ylmsz
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/gdbm-1.19-5p64xn325sx2eozu5ymsel2eleal745m
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/perl-5.34.0-ooiqj7n7xe5qhew7r37t4nl3yusqnuqu
[+] /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/openssl-1.1.1m-vfpiufs3nwu6rjqi3vewwghdoe6urrhm
==> Installing python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf
==> No binary for python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf found: installing from source
==> Using cached archive: /home/wdconinc/.spack/cache/_source-cache/archive/2c/2cc7b67c1f3f66c571acc42479cdf691d8ed6b47bee12c9b68430413a17a44ea.tgz
==> Applied patch /home/wdconinc/git/spack/var/spack/repos/builtin/packages/python/python-3.7.4+-distutils-C++.patch
==> Applied patch /home/wdconinc/git/spack/var/spack/repos/builtin/packages/python/python-3.7.4+-distutils-C++-testsuite.patch
==> Applied patch /home/wdconinc/git/spack/var/spack/repos/builtin/packages/python/tkinter-3.8.patch
==> Ran patch() for python
==> python: Executing phase: 'autoreconf'
==> python: Executing phase: 'configure'
==> python: Executing phase: 'build'
==> python: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j8' 'install'

2 errors found in build log:
     9716    Requirement already satisfied: pip in /usr/lib/python3/dist-packages (20.3.4)
     9717    Processing /tmp/tmpw31qfvag/pip-21.2.4-py3-none-any.whl
     9718    Installing collected packages: setuptools, pip
     9719      Attempting uninstall: setuptools
     9720        Found existing installation: setuptools 52.0.0
     9721        Uninstalling setuptools-52.0.0:
  >> 9722    ERROR: Could not install packages due to an OSError: [Errno 13] Permission denied: '__init__.py'
     9723    Consider using the `--user` option or check the permissions.
     9724    
     9725    Traceback (most recent call last):
     9726      File "/home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-src/Lib/runpy.py", line 197, in _run_module_as_main
     9727        return _run_code(code, main_globals, None,
     9728      File "/home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-src/Lib/runpy.py", line 87, in _run_code

     ...

     9735        return _run_pip(args + [p[0] for p in _PROJECTS], additional_paths)
     9736      File "/home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-src/Lib/ensurepip/__init__.py", line 34, in _run_pip
     9737        return subprocess.run([sys.executable, "-c", code], check=True).returncode
     9738      File "/home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-src/Lib/subprocess.py", line 528, in run
     9739        raise CalledProcessError(retcode, process.args,
     9740    subprocess.CalledProcessError: Command '['/home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-src/spack-build/python', '-c', '\nimport runpy\nimport sys\nsys.path = [\'/tmp/tmpw31qfv
             ag/setuptools-58.1.0-py3-none-any.whl\', \'/tmp/tmpw31qfvag/pip-21.2.4-py3-none-any.whl\'] + sys.path\nsys.argv[1:] = [\'install\', \'--no-cache-dir\', \'--no-index\', \'--find-links\', \'/tmp/tmpw31qfvag\', \'--root\', \'
             /\', \'--upgrade\', \'setuptools\', \'pip\']\nrunpy.run_module("pip", run_name="__main__", alter_sys=True)\n']' returned non-zero exit status 1.
  >> 9741    make: *** [Makefile:1264: install] Error 1

See build log for details:
  /home/wdconinc/.spack/stage/spack-stage-python-3.9.9-4wwfvrsagjcov5amcbqr5dwjofoxr2sf/spack-build-out.txt
```

The new `ensurepip` flag causes the python to find the system-installed `setuptools` in `/usr/lib/python3/dist-packages` (`Requirement already satisfied: pip in /usr/lib/python3/dist-packages (20.3.4)`).

The underlying issue is that `ensurepip` enables the configure flag `--with-ensurepip` which by default is taking the value `upgrade`. Setting this flag to `install` resolves the issue:
```diff
diff --git a/var/spack/repos/builtin/packages/python/package.py b/var/spack/repos/builtin/packages/python/package.py
index f8625b3505..5c95bf9045 100644
--- a/var/spack/repos/builtin/packages/python/package.py
+++ b/var/spack/repos/builtin/packages/python/package.py
@@ -492,7 +492,7 @@ def configure_args(self):
 
         if spec.satisfies('@2.7.9:2,3.4:'):
             if '+ensurepip' in spec:
-                config_args.append('--with-ensurepip')
+                config_args.append('--with-ensurepip=install')
             else:
                 config_args.append('--without-ensurepip')
 
```

However, this still will result in ensurepip finding `setuptools` and `pip` in `/usr/lib/python3/dist-packages`, so maybe this isn't quite a fix.

### Information on your system

* **Spack:** 0.17.1-952-d7179e7e8f
* **Python:** 3.9.7
* **Platform:** linux-ubuntu21.10-skylake
* **Concretizer:** clingo


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/var/spack/repos/builtin/packages/python/package.py
+++ b/var/spack/repos/builtin/packages/python/package.py
@@ -492,7 +492,7 @@ class Python(AutotoolsPackage):
 
         if spec.satisfies('@2.7.9:2,3.4:'):
             if '+ensurepip' in spec:
-                config_args.append('--with-ensurepip')
+                config_args.append('--with-ensurepip=install')
             else:
                 config_args.append('--without-ensurepip')
 

